Creation of Server Settings Based On Environment Variables

deef0000dragon1 opened the pull request:
Using the server settings files generated from a clean start of the container, This commit adds 3 template json files, and one script file to resolve #299

This script file uses 4 functions to validate different parameter inputs (string, boolean, number, or empty) and makes sure that the input from the environment variable matches the required format. It also makes sure that there is a value to replace it with. If either of these cases is false, the default value is used. 

Each file is created from the template in it's own function to allow for addition of or changing of keys in the future.

The current script is complete for the basic server settings, but not for the map settings or map-gen settings. The template keys have not yet been inserted into the template json file.

This is a request for review, to confirm that the required standards of error-proof and maintainable are met, and to discuss possible improvements.

SuperSandro2000 commented:
Pipeline will be green if you don't merge master into master.

SuperSandro2000 commented on 0.17/files/create-settings.sh:
```diff
@@ -0,0 +1,126 @@
+	val=${!3}
+
+	#Test if there was a value that was set.
+	if [ -z "$val" ]
```
Can you use ``[[`` everywhere?

SuperSandro2000 commented on 0.17/files/create-settings.sh:
```diff
@@ -0,0 +1,126 @@
+}
+
+mapSettings () {
+	echo Creating map generation settings file
```
Don't think this needs to be a function. This can be written at the bottom.

deef0000dragon1 replied:
The goal of making them functions is twofold. First, it separates the individual file template sections explicitly (I just havent filled the others out yet). Second, because they are called later in the program, they are parsed and validated before being run.

SuperSandro2000 replied:
Oh, I thought that this is done. Let me know when you filled those in.

SuperSandro2000 commented on 0.17/files/create-settings.sh:
```diff
@@ -0,0 +1,126 @@
+serverSettings(){
+	echo Creating server settings file
+	TEMPLATE_FILE=server-settings-template.json
+	updateTemplate templateServerName "my-server" TEMPLATE_SERVER_NAME
```
I don't know if it would look better but maybe we can format this tab seperated that is is more like a table.

deef0000dragon1 commented on 0.17/files/create-settings.sh:
```diff
@@ -0,0 +1,126 @@
+
+
+
+# Template String
```
Function Comments. Will clarify and format those.

deef0000dragon1 commented:
On the master to master, would it be best if I closed this and opened a new one? and would I merge a different branch on my end, or into a different branch on the factoriotools end?

SuperSandro2000 commented:
> On the master to master, would it be best if I closed this and opened a new one?

Don't know right now if you can change it. If yes change it if no just leave it like it is.

> and would I merge a different branch on my end, or into a different branch on the factoriotools end?

You would merge a different branch on your end into factoriotools master.
Take a look at https://github.com/factoriotools/factorio-docker/pull/267

SuperSandro2000 commented:
Looking great!

deef0000dragon1 commented:
So, unless you see anything that needs to be changed, I will make the necessary changes to the docker-entrypoint.

SuperSandro2000 commented on 0.17/files/map-gen-settings-template.json:
```diff
@@ -0,0 +1,70 @@
+		"richness": templateGenCliffRichness
+	},
+	"property_expression_names": {
+		"elevation": "templateGenExpressionElevation",
```
Did you miss some ``"`` here?

deef0000dragon1 replied:
those are in quotes in the example. so I left them in quotes. It was a bit of a strange section in the file to be honest.

SuperSandro2000 commented:
BTW How do you plan to update the templates in the future?

deef0000dragon1 commented:
By hand comparison. there shouldn't be too many, if ANY changes between minor versions at this point and few between major versions, unless they do a full settings refactor or something.

deef0000dragon1 commented:
Right then, documentation created/updated, and control variables added.

The new method will not run unless GENERATE_SETTINGS_FILES is true, otherwise, the old method(copy from /opt/factorio/data) is used. 

The value of FORCE_GENERATE_SETTINGS_FILES  will determine if the settings are generated and replaced if the files exist already. Unless true, the generation will not run if the files exist already. This allows for the override of settings if needed by restarting the container.

Unless you see anything that needs to be added, changed, or removed, it should be good to merge.
I would like if someone else could sanity check me and also test it, but it should be good.

SuperSandro2000 commented on 0.17/files/create-settings.sh:
```diff
@@ -232,23 +244,42 @@ serverSettings(){
+	serverSettings
+	mapSettings
 	mapGenSettings
+	rm -f "$CONFIG/server-settings.json"
```
I don't think it is a good idea to just delete the old configs.

deef0000dragon1 replied:
Files that exist will now get moved to .json.YYYY.MM.DD.HH.MM.SS

SuperSandro2000 replied:
I think this is a good solution to not loose any old settings. :+1:

deef0000dragon1 commented:
Im going to add a commit to add this to the 0.18 folder as well, and will try to test as in depth as I can this weekend.

deef0000dragon1 commented:
Closing for #304 . New branch name created accidentally. Running with it to allow for the checks to (hopefully) pass.
